skbio/diversity/beta/_unifrac.py

```diff
@@ -189,44 +199,304 @@ def weighted_unifrac(u_counts, v_counts, otu_ids, tree, normalized=False,
        73, 1576-1585 (2007).
 
     .. [2] Lozupone, C., Lladser, M. E., Knights, D., Stombaugh, J. & Knight,
-        R. UniFrac: an effective distance metric for microbial community
-        comparison. ISME J. 5, 169-172 (2011).
+       R. UniFrac: an effective distance metric for microbial community
+       comparison. ISME J. 5, 169-172 (2011).
 
     """
     if validate:
-        _validate(u_counts=u_counts, v_counts=v_counts,
-                  otu_ids=otu_ids, tree=tree)
-    u_obs_otu_counts = _observed_otu_counts(u_counts, otu_ids)
-    u_total_count = sum(u_counts)
-    v_obs_otu_counts = _observed_otu_counts(v_counts, otu_ids)
-    v_total_count = sum(v_counts)
-    u_obs_nodes = tree.observed_node_counts(u_obs_otu_counts)
-    v_obs_nodes = tree.observed_node_counts(v_obs_otu_counts)
-    uv_obs_nodes = set(u_obs_nodes) | set(v_obs_nodes)
-    if len(uv_obs_nodes) == 0:
-        # boundary case where both communities have no members
+        _validate(u_counts, v_counts, otu_ids, tree)
+
+    u_node_counts, v_node_counts, u_total_count, v_total_count, tree_index =\
+        _setup_single_unifrac(u_counts, v_counts, otu_ids, tree, validate,
+                              normalized=normalized, unweighted=False)
+    branch_lengths = tree_index['length']
+
+    if normalized:
+        tip_indices = _get_tip_indices(tree_index)
+        node_to_root_distances = _tip_distances(branch_lengths, tree,
```

Comment:
This calculation is being done pairwise, however it is expensive and should only be done one during full setup. Or is this method only expected to be called for a single pairwise op (e.g., passed as a callable to `beta_diversity`)? If so, it would be helpful to include a mention of this in the doc string